### ISSUE TITLE
How to make a plot with counts/bin on y axis and wavelength on x axis?

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#983]
The title says it all...
`set_analysis(9, "wave", "counts", factor=0)` gives counts/angstrom on the y-axis, the same with factor=1 gives `counts/Ang * wavelength^1 `on the y axis, but that's not the same as counts/bin, the y values are > 4000, while the number of counts is ~10 per bin. On the other hand, `set_analysis(9, "bin", "counts")`, gives me the correct y values, but not the x axis is in channels, not in wavelength.

I can get all the info out, but it's far more complicated than it should be:
```
set_analysis(9, "wave", "counts")
counts = get_data(9).get_dep(filter=True)  
wave = get_data_plot(9).x
from matplotlib import pyplot as plt
plt.plot(wave, counts)
```

Am I overlooking something obvious?

**Comment by DougBurke** [comment]
By counts/bin do you mean counts / bin-width where bin-width is in PI channels?

We have some long-standing discussions/attempts at work about this (e.g. #39 and #86) but I don't think we have come to a sensibe conclusion. I personally feel we should strip out the "plot" options from `set_analysis` and come up with a new scheme, but maybe we can come up changes (and probably new) arguments to `set_analysis`: I find the combination of `rate` and `factor` to be confusing.

Also note that you can (with the 4.12.2 code that's coming along) for PHA data say

    dplot = get_data_plot(9)

then

    dplot.xlo / dplot.xhi 

will get you the low and high edges of the bins.

**Comment by hamogu** [comment]
Yes, I see that those long-standing issues are similar or the same as this one. I'll leave this one open though, because it has a concrete question in it.

I want to plot count/bin. If, after grouping there is a bin that is three channels wide with a sum of5 counts it in, I want the y axis to show "5" (not 5/3. That's just silly for anyone who knows about Poisson statistics.)
The purpose of a plot like this is not so much physical interpretation like counts / keV, but in this way you can make a plot where you don't have to plot error bars to see how significant any error is. We often do that in grating spectroscopy so see emission lines. Any bin that's high, is a significant line. For that you need to know how many counts above continuum because it's really a question of signal-to-noise, not so much of how many erg of energy you get.

**Comment by DougBurke** [comment]
Keeping this open makes sense since one of my concerns with us looking at the existing tickets is we weren't really answering the question "What do we really want", more "what can we plot", so use cases are :+1:

**Comment by anetasie** [comment]
@hamogu I agree the plots should show 5 cts, not 5/3. This is an ongoing debate :-)

**Comment by DougBurke** [comment]
So, for a PHA dataset we have the following Y label for `plot_data` with `set_analysis` settings:

quantity | type | grouped | ylabel
-------- | ---- | ------- | ------
energy | rate | no | counts/sec/kev
energy | counts | no | counts
energy | rate | yes | counts/sec/kev
energy | counts | yes | counts/kev

I think we may want another flag to say whether we normalize by the bin width or not, since the table above doesn't make sense to me. I also need to dig into #86 a bit to finally put that to bed, as it may or may not be related.

**Comment by DougBurke** [comment]
I think https://github.com/sherpa/sherpa/blob/1a511737efa7b96669ac9a5f68af3d298f344758/sherpa/astro/data.py#L3121 is an interesting place to look at,

**Comment by hamogu** [comment]
How do I set the "grouped" flag? Do you mean group / ungroup? Because I do really want counts / channel. That's not a terribly physical unit (i.e. "per Ang" might sound better) but it's visually useful ind that practitioners immediately see he S/N per bin.

**Comment by hamogu** [comment]
Here is an example what I have in mind:
![image](https://user-images.githubusercontent.com/498688/106067718-41986280-60cd-11eb-8da2-29b3b0185e52.png)
From https://www.aanda.org/articles/aa/full/2008/45/aa10054-08/aa10054-08.html

**Comment by DougBurke** [comment]
you could try 

  set_grouping(np.ones(nchans))
  group()

which should tell Sherpa to group the data, it's just that each group has a single channel. It may work.

So, here's the result of `plot('data', 1, 'data', 2)` which was

    load_pha(1, infile)
    load_pha(2, infile)
    load_arf(1, arffile)
    load_arf(2, arffile)
    set_analysis('wave', type='counts')
    ungroup(1)
    ungroup(2)
    set_grouping(2, np.ones(16384))
    group(2)

Note that this divides by the bin width in Angstrom.

![bins](https://user-images.githubusercontent.com/224638/106073780-b2914780-60d8-11eb-86b0-48045a1f217a.png)


Hmm, that's a good "idempotency" test - you should get the same results for a fit or plot (except for the very issue we're talking about here) whether we have a everything-is-grouped-to-one-channel vs no grouping.

**Comment by DougBurke** [comment]
But I'm not quite sure that's what you want...

**Comment by hamogu** [comment]
So, I just have to make sure it's not binned to solve my original question. Problem solved; I only wonder why I did not find that. Either it's not obvious or I have to go and read the docs again.

**Comment by DougBurke** [comment]
I think there's still reasons to leave this open, as you may want to divide by the number of channels in  the group even with analysis=wave.

Conceptually we want to be able to apply a user-supplied function to go from <data, stat, maybe-more-but-I-can't-think-of-one> to <xvals, y, dy> for plotting. Then we have  a bunch of pre-canned routines which we can easily tweak (eg change from rate to counts) so you don't need to specify a function, and then there may be a way to fall-back to user-wants-to-do-it-manually.

The data class hierarchy has a `to_plot` method, so we have an API, and this is in fact how analysis/rate/factor are applied. I just think that at the moment we do not have a good sense of what we want to plot, for example

  -  I don't think the table above makes much sense logically https://github.com/sherpa/sherpa/issues/983#issuecomment-768598976
  - #1021 describes how we display models differently for plot_model and plot_fit and, while I now understand it, it either needs better documentation or a way to say "group or don't group the plot data" (ideally the latter, and not just for model)
  - <argh, I had another item to add, but I have forgotten>
